runtime/v2/runc/task/service.go

```diff
@@ -509,56 +545,72 @@ func (s *service) Stats(ctx context.Context, r *taskAPI.StatsRequest) (*taskAPI.
 
 func (s *service) processExits() {
 	for e := range s.ec {
-		s.checkProcesses(e)
+		s.lifecycleMu.Lock()
+		if cps, ok := s.running[e.Pid]; ok {
+			delete(s.running, e.Pid)
+			s.lifecycleMu.Unlock()
+			for _, cp := range cps {
+				s.handleProcessExit(e, cp.Container, cp.Process)
+			}
+		} else {
+			s.earlyExits[e.Pid] = append(s.earlyExits[e.Pid], e)
+			s.lifecycleMu.Unlock()
+		}
 	}
 }
 
-func (s *service) send(evt interface{}) {
-	s.events <- evt
+func (s *service) earlyExitsGC() {
+	for {
+		select {
```

Comment:
This could be combined with the `processExits` loop and since that single go routine would be the only one adding to earlyExits, you would only need to set the timer when earlyExits is not empty.